### PR TITLE
Add requeue method for datapoints.

### PR DESCRIPTION
**Pull request by brianlball**
Add spot instance polling scripts and trigger requeue

### CODE REVIEW

**Comment by brianlball** [comment]
https://github.com/NREL/OpenStudio-server/issues/754

**Comment by brianlball** [comment]
This test will trigger the /requeue api after a few seconds to requeue a running datapoint
https://github.com/NREL/OpenStudio-server/blob/requeue/server/spec/features/docker_stack_requeue_spec.rb#L185

**Comment by brianlball** [comment]
the AWS metadata spot instance termination scripts were mock tested [here](https://github.com/NREL/OpenStudio-server/blob/requeue/local_setup_scripts/win64/docker-compose-mock.yml#L110) with [amazon-ec2-metadata-mock](https://github.com/aws/amazon-ec2-metadata-mock)

File: docker/server/aws_imdsv2.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+# Metadata service URL variable
```
Comment (nllong) [inline docker/server/aws_imdsv2.sh:12]: can you add a paragraph at the top of this script that explains what it is doing and what the user would need to set?

File: docker/server/run-server-tests.sh [hunk 2]
```diff
@@ -24,6 +24,7 @@ done
 cd /opt/openstudio/server && bundle exec rspec spec/features/docker_stack_custom_gems.rb; (( exit_status = exit_status || $? ))
 cd /opt/openstudio/server && bundle exec rspec spec/features/docker_stack_test_apis_spec.rb; (( exit_status = exit_status || $? ))
 cd /opt/openstudio/server && bundle exec rspec spec/features/docker_stack_algo_spec.rb; (( exit_status = exit_status || $? ))
+cd /opt/openstudio/server && bundle exec rspec spec/features/docker_stack_requeue_spec.rb; (( exit_status = exit_status || $? ))
```
Comment (nllong) [inline docker/server/run-server-tests.sh:27]: sweet

File: server/spec/features/docker_stack_requeue_spec.rb [hunk 3]
```diff
@@ -0,0 +1,259 @@
+# *******************************************************************************
+# OpenStudio(R), Copyright (c) Alliance for Sustainable Energy, LLC.
+# See also https://openstudio.net/license
+# *******************************************************************************
+
+#################################################################################
+# To Run this test manually:
+#
+#   start a server stack with /spec added and ssh into the Web container
```
Comment (nllong) [inline server/spec/features/docker_stack_requeue_spec.rb:9]: this is cool

**Review by nllong — APPROVED** [review]
looks like some great functionality being added here!

**Comment by brianlball** [comment]
also, for manual mock testing, in a worker, download the ec2-meta-mock and have it interrupt in 60seconds:
``root@b4838250bc28:/opt/openstudio/server# curl -Lo ec2-metadata-mock https://github.com/aws/amazon-ec2-metadata-mock/releases/download/v1.11.2/ec2-metadata-mock-`uname | tr '[:upper:]' '[:
lower:]'`-amd64``
`root@b4838250bc28:/opt/openstudio/server# chmod +x ec2-metadata-mock`
`root@b4838250bc28:/opt/openstudio/server# ./ec2-metadata-mock -d 60 spot -I`

you can tail the aws_imds.log and you should see the script refresh the token once its avail, then trigger the requeue for the worker PID on this node, then send the KILL sigterm to resque, which allows for the resque worker to finish the job if possible, and not process any more jobs.  TERM immediatly kills the worker, which is a commented out option, but I dont think we want to do that.

`root@b4838250bc28:/opt/openstudio/server# tail -f log/aws_imds.log`
`Token Refresh Failed`
`No interruption detected. Continuing to monitor.`
`Refreshing Authentication Token`
`No interruption detected. Continuing to monitor.`
`Instance interruption notice received. Requeuing datapoint.`
`REQUEUE: getting workers`
`RESQUEUE: found workers: ["b4838250bc28:46:simulations", "cbbd1fcca759:44:background,analyses", "bf71c27486f8:50:a
nalysis_wrappers", "04fc93d09ce0:46:simulations"]`
`RESQUEUE: getting worker: b4838250bc28:46:simulations for b4838250bc28 in queue: simulations`
`RESQUEUE: working_on: {"queue":"simulations","run_at":"2024-04-10T14:30:33Z","payload":{"class":"ResqueJobs::RunSi
mulateDataPoint","args":["28d6a027-8b56-43ec-9350-7ec3bfe14f9d"]}}`
`RESQUEUE: found local worker`
`RESQUEUE: job_data: {"queue"=>"simulations", "run_at"=>"2024-04-10T14:30:33Z", "payload"=>{"class"=>"ResqueJobs::R
unSimulateDataPoint", "args"=>["28d6a027-8b56-43ec-9350-7ec3bfe14f9d"]}}`
`REQUEUE: Worker b4838250bc28:46:simulations on this node is processing a job of class ResqueJobs::RunSimulateDataP
oint with args 28d6a027-8b56-43ec-9350-7ec3bfe14f9d`
`REQUEUE: calling requeue_datapoint`
REQUEUE: URI http://web/data_points/28d6a027-8b56-43ec-9350-7ec3bfe14f9d/requeue`
`REQUEUE: Requeued datapoint with UUID 28d6a027-8b56-43ec-9350-7ec3bfe14f9d. Response: 302 Found`
`REQUEUE: PID: 46`
`REQUEUE: Sending KILL signal to worker with PID 46.`
`REQUEUE: Sent QUIT signal to worker with PID 46.`

**Pull request merged**